### ISSUE TITLE
SignalR npm package not able to be bundled using `esbuild`

### ISSUE THREAD

**Issue by hrueger** [dotnet/aspnetcore#47674]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

Hi,
I'm using the npm package `@microsoft/signalr` with version `7.0.4`. I'm using `esbuild` for bundeling, however, it does not correctly resolve some dependencies, like `ws`.

This is due [those hacks](https://github.com/search?q=repo%3Adotnet%2Faspnetcore%20__webpack_require__&type=code):
```ts
// In order to ignore the dynamic require in webpack builds we need to do this magic
// @ts-ignore: TS doesn't know about these names
const requireFunc = typeof __webpack_require__ === "function" ? __non_webpack_require__ : require;
webSocketModule = requireFunc("ws");
eventSourceModule = requireFunc("eventsource");
```

I don't think this is an issue of `esbuild`, since it believes that this is a dynamic require. Can we somehow fix this?

I confirmed that this is the issue, by patching the lib to use plain `require` instead of the `requireFunc`, then everything works correctly.

### Expected Behavior

`ws` is resolved correctly when using the `bundle` option of `esbuild`.

### Steps To Reproduce

https://github.com/hrueger/signalr-esbuild-issue

1. `yarn install`
1. `yarn build`
1. See `__webpack_require__` is still in `dist/bundle.js` instead of the `ws` module embedded in the bundle.

### .NET Version

n/a

**Comment by BrennanConroy** [comment]
Looks like an esbuild issue
https://github.com/evanw/esbuild/issues/1921

**Comment by hrueger** [comment]
Are you sure? I don't think so, to be honest. This is not about esm / cjs, but instead the workaround which is implemented for webpack here, breaks other bundlers (at least esbuild). To those, the call looks like a dynamic require, although it is actually not dynamic. Do you know, what I mean?

**Comment by shsuman** [comment]
We are facing the same issue with 5.0.17. The workaround was to replace the **\_\_non\_webpack\_require\_\_** with **require** using webpack plugin.

```javascript
{
      test: /@microsoft.signalr.dist.cjs.(HttpConnection|FetchHttpClient)\.js/,
      loader: 'string-replace-loader',
      options: {
             search: /__non_webpack_require__/g,
             replace: 'require',
             strict: true
       }
}
``` 
The question I want to ask is why is **\_\_non_webpack_require\_\_** is being used because after searching online, it seems like it is not meant to be used in production code and no other package uses it.

**Comment by hrueger** [comment]
Thanks!

**Comment by BrennanConroy** [comment]
Unfortunately this issue will need to be reopened. The fix we made broke other important scenarios so we'll need to find a different way to fix this. Starting with 8.0.7 this esbuild issue will start happening again.

**Comment by rahul-sharma-uipath** [comment]
@BrennanConroy did it break multiple scenarios or just the esbuild thing?

**Comment by sschultze** [comment]
For bundling with `rollup` (for Electron Main), I also have to use an ugly workaround involving the `replace` plugin. Needless to say that this is very unsatisfactory.

```js
replace({
    preventAssignment: true,
    include: [
        './node_modules/@microsoft/signalr/dist/**/*'
    ],
    values: {
        // @microsoft/signalr 8.0.7
        '/*#__PURE__*/': '',
        'new (requireFunc("tough-cookie")).CookieJar()': 'undefined',
        'this._fetchType = requireFunc("fetch-cookie")(this._fetchType, this._jar);': JSON.stringify(''),
        'requireFunc("ws")': 'undefined',
        'requireFunc("eventsource")': 'undefined'
    },
    delimiters: ['', '']
}),
```

In addition, the private `WebSocket` and `EventSource` properties have to be set in `IHttpConnectionOptions` manually. And this workaround wouldn't work with cookies.

**Comment by bigcakes** [comment]
Any updates on this issue?

**Comment by nima-ap** [comment]
This is causing issues with webpack (5.98.0) as well when the application bundle targets web, specifically in development mode where more aggressive tree-shaking and minimization is disabled.

import { createRequire as __WEBPACK_EXTERNAL_createRequire } from "node:module";

Is included in the bundle which does not work in browser environments without a hacky workaround in the webpack config to exclude the non-browser code:

```
resolve: {
    alias: {
        '@microsoft/signalr': path.resolve(__dirname, ...toRootNodeModules, '@microsoft', 'signalr', 'dist', 'browser', 'signalr.js'),
    },
    // Exclude Node core modules that eventsource (dep of signalR) is trying to require
    fallback: {
        http: false,
        https: false,
    },
},
```

**Comment by nulladdict** [comment]
This issue is part of the larger problem with the published bundle. Despite providing `dist/browser/signalr.js` and `dist/webworker/signalr.js` builds there are still hacks and runtime feature detection inside the bundled code.

Browser build should never even include the `requireFunc("tough-cookie")`, `requireFunc("node-fetch")`, `requireFunc("ws")` and other node.js specific bits.

The code hidden behind `requireFunc` hacks is a major issue for bundlers:
- Some, like esbuild, fail to build completely
- Some, like vite, ship the dead code straight to browser, but eliminate the actual dependency from the bundle
- Some might ship both the dead code and the unused dependencies

There's no need for webpack-specific build hacks in 2025. The proper way to do this is:
- Remove the `requireFunc` hacks
- Remove the runtime detection through `Platform.is` hacks
- Instead use some form of build-time code elimination and provide separate bundles for different environments
  - These bundles can be shipped as a single package with "exports" conditions or as separate packages (e.g. `@microsoft/signalr-node`, `@microsoft/signalr-browser`, `@microsoft/signalr-webworker`, etc.)

If you don't want to bother with proper bundling you can also force consumers to provide the required dependencies in constructors or through polyfills:
- `FetchHttpClient` could accept `fetch` with cookie support and jar if needed
- `AbortController` can be an external requirement and polyfilled by the consumer if needed
- `HttpConnection` could accept `WebSocket` and `EventSource` constructors, or rely on them existing globally

Basically there's no need for this runtime extravaganza. All the dependency differences can be turned either into explicit or implicit environment requirements.

**Comment by manbearwiz** [comment]
In case this is helpful to anyone else, heres the replace config I came up with to create a bundle for `7.0.14` that works in a modern browser with all of the node stuff shaken out

```ts
    plugins: [
      // Replace SignalR's platform detection with browser-specific values
      // This eliminates Node.js-specific code paths and optimizes the bundle for browser use
      replace({
        values: {
          "Platform.isNode": JSON.stringify(false),
          "Platform.isBrowser": JSON.stringify(true),
          "Platform.isReactNative": JSON.stringify(false),
          "typeof fetch": JSON.stringify("function"),
          "typeof require": JSON.stringify("undefined"),
          "typeof globalThis": JSON.stringify("object"),
          "typeof AbortController": JSON.stringify("function"),
        },
        preventAssignment: true,
        sourcemap: true,
      }),
      terser(),
    ],
```

**Comment by CaringDev** [comment]
With more and more applications / frameworks moving to newer tools, this is becoming increasingly urgent.
A fix before .NET 11 would be greatly appreciated.

**Comment by AntoineFra** [comment]
I put a comment on a linked issue with MF, if it helps anyone.
[Here](https://github.com/dotnet/aspnetcore/issues/57145#issuecomment-4798905597)
